### DISCUSSION TITLE
Server Side Version Counters -- should I use?

### DISCUSSION THREAD

**Discussion by davidroeca** [sqlalchemy/sqlalchemy#11533] [Usage Questions] [answered]
I've been browsing the docs on version counters and came across [Server Side Version Counters](https://docs.sqlalchemy.org/en/20/orm/versioning.html#server-side-version-counters).

They seem really convenient, because they don't require an additional db migration or custom code, since I'm using PostgreSQL already.

I'd assume one benefit of server side version counters is that they avoid the warning that mentions:

> Because the versioning feature relies upon comparison of the in memory record of an object, the feature only applies to the [Session.flush()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.flush) process, where the ORM flushes individual in-memory rows to the database. It does not take effect when performing a multirow UPDATE or DELETE

But please let me know if that assumption is incorrect.

Then I also saw the note at the bottom:

> It is strongly recommended that server side version counters only be used when absolutely necessary ...

I'm using PostgreSQL, which supports RETURNING, but I'm curious to hear other reasons that might guide me away from this approach.

**Comment by zzzeek** [comment, marked as answer]
I think that note about "strongly recommended to avoid" might be a little dated at this point and probably referred more to other. more complicated ways of doing this like using triggers or other awkward things like that, PG's "xmin" is obviously pretty well suited to this, but you are still coding to something fairly internal to PostgreSQL's implementation.   So obviously it's going to lock you into PostgreSQL (as well as whatever internal changes they might make which would impact how "xmin" works), and prevent things like say having your test suite run against SQLite, but I guess that warning was mostly a "hey please don't get into this if you dont know what you're doing" kind of thing, and also, our RETURNING support for defaults like these has come a long way since that doc was written.    So it's probably fine though I'd want to research some real world shops that use xmin and confirm this is an industry standard approach coming from the PG side.

**Reply by CaselIT** [reply]
do you plan to update that part of the docs?

**Reply by zzzeek** [reply]
we should just take that note out, those notes are usually to prevent novice users coming to us with problems they shouldnt have but i dont think this one is helping us

**Reply by davidroeca** [reply]
Thanks for the quick reply -- this makes sense to me. I'll keep these trade offs in mind as I work on my proof of concept
